AK/Vector.h

```diff
@@ -709,6 +709,18 @@ requires(!IsRvalueReference<T>) class Vector {
         MUST(try_resize_and_keep_capacity(new_size));
     }
 
+    void shrink_to_fit()
+    {
+        if (size() == capacity())
+            return;
+        Vector new_vector;
+        new_vector.ensure_capacity(size());
+        for (auto& element : *this) {
+            new_vector.unchecked_append(move(element));
+        }
+        *this = move(new_vector);
```

Comment:
Could everything after the early return just be:
```c++
Vector new_vector { *this };
*this = move(new_vector);
```
?
This would use `TypedTransfer`, which seems "optimized" for trivial types

Reply:
Though I wonder if instead of creating a temporary copy, if we could do something like:
```c++
if (m_outline_buffer) {
    kfree_sized(m_outline_buffer + (m_size * sizeof(StorageType)), (m_capacity - m_size) * sizeof(StorageType));
    m_capacity = m_size;
}
```

(I ask that having no idea if our kmalloc supports freeing tail memory)

Reply:
> Could everything after the early return just be:
> 
> ```c++
> Vector new_vector { *this };
> *this = move(new_vector);
> ```
> 
> ? This would use `TypedTransfer`, which seems "optimized" for trivial types

This would not free any excess capacity as `Vector::operator=(Vector&&)` just transfers ownership of the existing buffer from one vector to another.


> Though I wonder if instead of creating a temporary copy, if we could do something like:
> 
> ```c++
> if (m_outline_buffer) {
>     kfree_sized(m_outline_buffer + (m_size * sizeof(StorageType)), (m_capacity - m_size) * sizeof(StorageType));
>     m_capacity = m_size;
> }
> ```
> 
> (I ask that having no idea if our kmalloc supports freeing tail memory)

We don't support freeing tail memory, so that won't work :)